src/python/pants/util/objects.py

```diff
@@ -266,59 +264,41 @@ def __init__(self, type_name, msg, *args, **kwargs):
       type_name, formatted_msg, *args, **kwargs)
 
 
+# TODO: make these members of the `TypeConstraint` class!
 class TypeConstraintError(TypeError):
   """Indicates a :class:`TypeConstraint` violation."""
 
 
 class TypeConstraint(AbstractClass):
   """Represents a type constraint.
 
-  Not intended for direct use; instead, use one of :class:`SuperclassesOf`, :class:`Exact` or
+  Not intended for direct use; instead, use one of :class:`SuperclassesOf`, :class:`Exactly` or
   :class:`SubclassesOf`.
   """
 
-  def __init__(self, *types, **kwargs):
+  def __init__(self, variance_symbol, description):
     """Creates a type constraint centered around the given types.
 
     The type constraint is satisfied as a whole if satisfied for at least one of the given types.
 
     :param type *types: The focus of this type constraint.
     :param str description: A description for this constraint if the list of types is too long.
     """
-    if not types:
-      raise ValueError('Must supply at least one type')
-    if any(not isinstance(t, type) for t in types):
-      raise TypeError('Supplied types must be types. {!r}'.format(types))
-
-    # NB: `types` is converted to tuple here because self.types's docstring says
-    # it returns a tuple. Does it matter what type this field is?
-    self._types = tuple(types)
-    self._desc = kwargs.get('description', None)
-
-  @property
-  def types(self):
-    """Return the subject types of this type constraint.
-
-    :type: tuple of type
-    """
-    return self._types
-
-  def satisfied_by(self, obj):
-    """Return `True` if the given object satisfies this type constraint.
-
-    :rtype: bool
-    """
-    return self.satisfied_by_type(type(obj))
+    assert(variance_symbol)
+    self._variance_symbol = variance_symbol
+    self._description = description
 
   @abstractmethod
-  def satisfied_by_type(self, obj_type):
+  def satisfied_by(self, obj):
     """Return `True` if the given object satisfies this type constraint.
 
     :rtype: bool
     """
 
   def validate_satisfied_by(self, obj):
-    """Return `obj` if the object satisfies this type constraint, or raise.
+    """Return some version of `obj` if the object satisfies this type constraint, or raise.
```

Comment:
What does "some version" mean?

Reply:
Nothing, after removing `wrapper_type`. I will remove this!